[8.x] Improvements to Str

thomas-alrek opened the pull request:
<!--
Please only send a pull request to branches which are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

I find myself using the `Str` class very much, as the various string functions in PHP can be "messy" to use. One thing that I often feel is missing when working with Laravel in my day job, is a consistent wrapper for the `explode` and `implode` methods, that also supports Collections.

This pull request adds two new static methods to `\Illuminate\Support\Str`.

**Str::split**
---
This method provides a proxy for `explode`, returning a `\Illuminate\Support\Collection` of `\Illuminate\Support\Stringable` objectss.

#### Example
```php
<?php

use Illuminate\Support\Str;

$str = 'foo,bar,baz';
$delimiter = ',';

$pieces = Str::split($str, $delimiter); // (Collection) => ['foo', 'bar', 'baz']
```

**Str::join**
---
This method provides a proxy for `implode`, taking an array or a `\Illuminate\Support\Contracts\Arrayable` and joins them back to a `\Illuminate\Support\Stringable` object.

#### Example
```php
<?php

use Illuminate\Support\Str;

$glue = ',';
$pieces = ['foo', 'bar', 'baz'];

$str = Str::join($pieces, $glue); // (Stringable) => 'foo,bar,baz'
```

thomas-alrek committed:
Add static constructor

thomas-alrek committed:
Add split and join static methods

thomas-alrek committed:
Fix StyleCI issue

thomas-alrek committed:
Add tests for split and join

thomas-alrek committed:
Add test for make

thomas-alrek committed:
Update tests

GrahamCampbell commented on src/Illuminate/Support/Str.php:
```diff
@@ -794,4 +795,39 @@ public static function createUuidsNormally()
+     *
+     * @param  string  $string
+     * @param  string  $delimiter
+     * @param  int  $limit
```
`int|null`

GrahamCampbell commented on src/Illuminate/Support/Str.php:
```diff
@@ -794,4 +795,39 @@ public static function createUuidsNormally()
+    public static function split($string, $delimiter = ',', $limit = null)
+    {
+        $parameters = array_filter([$delimiter, $string, $limit]);
+        $pieces = call_user_func_array('explode', $parameters);
```
`explode(...array_filter([$delimiter, $string, $limit]))`?

GrahamCampbell commented on src/Illuminate/Support/Stringable.php:
```diff
@@ -30,6 +30,18 @@ public function __construct($value = '')
+     * Create a new stringable instance.
+     *
+     * @param  string  $items
+     *
```
cs

derekmd commented on src/Illuminate/Support/Str.php:
```diff
@@ -794,4 +795,39 @@ public static function createUuidsNormally()
+        $pieces = call_user_func_array('explode', $parameters);
+
+        return Collection::make($pieces)
+            ->map([Stringable::class, 'make']);
```
This could be replaced with `->mapInto(Stringable::class)` which removes the need to add a `make()` method.

thomas-alrek committed:
Remove make method

thomas-alrek committed:
Refactor according to review feedback

VeryStrongFingers commented on src/Illuminate/Support/Str.php:
```diff
@@ -794,4 +795,38 @@ public static function createUuidsNormally()
+     */
+    public static function split($string, $delimiter = ',', $limit = null)
+    {
+        $pieces = explode(...array_filter([$delimiter, $string, $limit]));
```
You should allow `$delimiter` to be falsey - might be best to avoid `array_filter()`

thomas-alrek committed:
Handle empty strings, use PHP_INT_MAX as the default limit like explode does.

thomas-alrek committed:
Only filter elements with strlen of 0.

PHP's explode has a quirk where splitting an empty string returns an array with one empty element, instead of an empty array.

thomas-alrek committed:
Update tests

taylorotwell commented:
Why not just call them explode and implode?

thomas-alrek commented:
> Why not just call them explode and implode?

I thought about, but I wanted it to align with the similar methods in the Collection class, and I also find the names explode and implode very "confusing". In most other languages the similar methods are usually named split and join

taylorotwell commented:
Hmm, I think I'll just leave things be for now - Stringable already has an explode() and collection already has an implode.
